### DISCUSSION TITLE
Changing `otp_secret_encryption_key`

### DISCUSSION THREAD

**Discussion by pmichna** [devise-two-factor/devise-two-factor#219] [Q&A]
Hi, I have two questions.

1. Suppose the `otp_secret_encryption_key` leaks. What's the danger - how can an attacker harm the application having the key?

2. What's the best way to rotate they? Deactivate OTP requirement for all the users, change the `otp_secret_encryption_key` and make users enable 2FA again?

**Comment by shaicoleman** [comment]
1. The `otp_secret_encryption_key` is only useful in conjunction with the contents of the users table.
That said, if an attacker has managed to obtain the encryption key, there's a very good chance they'll be able to access the database.

2. The best approach would be to implement some kind of key versioning, e.g. by adding a field to the users table `otp_key_version` or by using a date field e.g. `otp_updated_at`, and then make the encryption key dynamic (e.g. everything after a certain date uses encryption key v2). Once the users authenticate as normal, including 2FA with the old encryption key, you force them to reset their password and set up 2FA again.

This isn't something that's supported by the gem, so you'll have to build that functionality yourself.

**Comment by shaicoleman** [comment]
Replacing `attr_encrypted` with `lockbox` https://github.com/ankane/lockbox seems like the best way to do it. The `lockbox` gem has built-in support for key rotation
